linera-chain/src/manager.rs

```diff
@@ -394,36 +398,45 @@ impl ChainManager {
         Ok(Outcome::Accept)
     }
 
+    /// If the validated block certificate is more recent, returns that certificate
+    /// so that the locked block can be updated.
+    pub fn validated_block_if_newer_than_locked(
+        &self,
+        validated_block_certificate: Option<LiteCertificate<'static>>,
+        executed_block: ExecutedBlock,
+    ) -> Option<Certificate> {
+        let lite_cert = validated_block_certificate?;
+        if self
+            .locked
+            .as_ref()
+            .map_or(true, |locked| locked.round < lite_cert.round)
+        {
+            let value = HashedCertificateValue::new_validated(executed_block);
+            return lite_cert.with_value(value);
```

Comment:
We're putting the `executed_block` inside an already-existing certificate which was created for possibly a different value, different block. The signatures on this certificate will not match.

 This whole `LiteCertificate::with_value` function is dangerous. I know it checks whether it's for the same hash but the idea of "recreating" a certificate feels bad.